### ISSUE TITLE
Results are not good 

### ISSUE THREAD

**Issue by MotiBaadror** [switchablenorms/DeepFashion_Try_On#41]
Below are the testing results on my data.

1. I have used pose_keypoints from coco_18 model and verified the keypoints by drawing on the image and are good for the below results. 
2. Generated labels with pretrained model o LIP dataset and modified the numbering accordingly. 
3. I am using test_color and test_edges same as test data provided.

The results are not completely satisfying. Any suggestions are welcome.

Thank you

![image](https://user-images.githubusercontent.com/28961913/92901839-736e9c80-f43e-11ea-80f4-22cdcc7ae256.png)
![image](https://user-images.githubusercontent.com/28961913/92902279-c34d6380-f43e-11ea-9ece-7caeac8daed7.png)

Little v shape in neck is same as query image
![image](https://user-images.githubusercontent.com/28961913/92902305-c9434480-f43e-11ea-91f1-5d143001327f.png)

Sleeves are same as query image
![image](https://user-images.githubusercontent.com/28961913/92902326-cfd1bc00-f43e-11ea-94d8-b417c21c7018.png)

Overlaying the back layer of t-shirt on front neck
![image](https://user-images.githubusercontent.com/28961913/92902379-dbbd7e00-f43e-11ea-83e4-96895e621768.png)

Left arms cloth is not overlaying
![image](https://user-images.githubusercontent.com/28961913/92902452-e9730380-f43e-11ea-8dd5-c5a5d2d5cd9e.png)

Cloth is same as query image
![image](https://user-images.githubusercontent.com/28961913/92906399-107f0480-f442-11ea-8b1a-f7b8a40f9e63.png)

**Comment by taliegemen** [comment]
This is a field that research is still being made, and the code will not give completely satisfactory results. But in your case, segmentation is badly done and thats why try-on module is not working as intended.

**Comment by Mushahid2521** [comment]
Will you please help me how you generated the results? I am running the code on Google Colab. I could run the test.py file successfully but I don't see the result in sample directory. It generate an image file with 5 images concatenated but I don't find the output.

**Comment by MotiBaadror** [comment]
Hi @taliegemen Thank you for your response. What do you suggesting? What segmentation technique I should use? Cloth edges I am using the same edges that are given in the test dataset. I have only changed the query image and label and keypoints for the same.

**Comment by Mushahid2521** [comment]
@MotiBaadror will you please share how you have generated the pose of a custom image.
Thanks a lot.

**Comment by MotiBaadror** [comment]
> @MotiBaadror will you please share how you have generated the pose of a custom image.
> Thanks a lot.
@Mushahid2521 , You can follow the [link](https://www.programmersought.com/article/3282857837/)

**Comment by Mushahid2521** [comment]
> > @MotiBaadror will you please share how you have generated the pose of a custom image.
> > Thanks a lot.
> > @Mushahid2521 , You can follow the [link](https://www.programmersought.com/article/3282857837/)

Thanks for the tutorial. I think they haven’t described how to output the predictions in json format like the way used in this project.       
Will you please help me on that?
Thanks.

**Comment by MotiBaadror** [comment]
> > > @MotiBaadror will you please share how you have generated the pose of a custom image.
> > > Thanks a lot.
> > > @Mushahid2521 , You can follow the [link](https://www.programmersought.com/article/3282857837/)
> 
> Thanks for the tutorial. I think they haven’t described how to output the predictions in json format like the way used in this project.
> Will you please help me on that?
> Thanks.

Use the below format, just modify the value of that dictionary key (pose_keypoints) rest will be the same. 
`{"version": 1.3, "people": [{"person_id": [-1], "pose_keypoints": [], "face_keypoints_2d": [], "hand_left_keypoints_2d": [], "hand_right_keypoints_2d": [], "pose_keypoints_3d": [], "face_keypoints_3d": [], "hand_left_keypoints_3d": [], "hand_right_keypoints_3d": []}]}`


Also you have to get confidence from predict function.

`       points = []
        conf=[]
        for idx in range(self.num_points):
            probMap = output[0, idx, :, :] # confidence map.

            # Find global maxima of the probMap.
            minVal, prob, minLoc, point = cv2.minMaxLoc(probMap)

            # Scale the point to fit on the original image
            x = (img_width * point[0]) / W
            y = (img_height * point[1]) / H

            if prob > self.threshold:
                points.append((int(x), int(y)))
                conf.append(prob)
            else:
                points.append(None)
                conf.append(0)

        return points,conf
`

**Comment by taliegemen** [comment]
@MotiBaadror CHIP_PGN method is the best atm as far as i know (since it segmentates neck section as well). But to use it directly you have to retrain the model with more amount of classes since PGN generates segmentation images with more classes.

**Comment by ecarist** [comment]
Will you please show the specific running commands during testing? thanks!

**Comment by ecarist** [comment]
> Below are the testing results on my data.
> 
>     1. I have used pose_keypoints from coco_18 model and verified the keypoints by drawing on the image and are good for the below results.
> 
>     2. Generated labels with pretrained model o LIP dataset and modified the numbering accordingly.
> 
>     3. I am using test_color and test_edges same as test data provided.
> 
> 
> The results are not completely satisfying. Any suggestions are welcome.
> 
> Thank you
> 
> ![image](https://user-images.githubusercontent.com/28961913/92901839-736e9c80-f43e-11ea-80f4-22cdcc7ae256.png)
> ![image](https://user-images.githubusercontent.com/28961913/92902279-c34d6380-f43e-11ea-9ece-7caeac8daed7.png)
> 
> Little v shape in neck is same as query image
> ![image](https://user-images.githubusercontent.com/28961913/92902305-c9434480-f43e-11ea-91f1-5d143001327f.png)
> 
> Sleeves are same as query image
> ![image](https://user-images.githubusercontent.com/28961913/92902326-cfd1bc00-f43e-11ea-94d8-b417c21c7018.png)
> 
> Overlaying the back layer of t-shirt on front neck
> ![image](https://user-images.githubusercontent.com/28961913/92902379-dbbd7e00-f43e-11ea-83e4-96895e621768.png)
> 
> Left arms cloth is not overlaying
> ![image](https://user-images.githubusercontent.com/28961913/92902452-e9730380-f43e-11ea-8dd5-c5a5d2d5cd9e.png)
> 
> Cloth is same as query image
> ![image](https://user-images.githubusercontent.com/28961913/92906399-107f0480-f442-11ea-8b1a-f7b8a40f9e63.png)

May I ask how to print the final result:try-on image?   how do you combine them in the last column of the picture？

**Comment by MotiBaadror** [comment]
> > Below are the testing results on my data.
> > ```
> > 1. I have used pose_keypoints from coco_18 model and verified the keypoints by drawing on the image and are good for the below results.
> > 
> > 2. Generated labels with pretrained model o LIP dataset and modified the numbering accordingly.
> > 
> > 3. I am using test_color and test_edges same as test data provided.
> > ```
> > 
> > 
> > The results are not completely satisfying. Any suggestions are welcome.
> > Thank you
> > ![image](https://user-images.githubusercontent.com/28961913/92901839-736e9c80-f43e-11ea-80f4-22cdcc7ae256.png)
> > ![image](https://user-images.githubusercontent.com/28961913/92902279-c34d6380-f43e-11ea-9ece-7caeac8daed7.png)
> > Little v shape in neck is same as query image
> > ![image](https://user-images.githubusercontent.com/28961913/92902305-c9434480-f43e-11ea-91f1-5d143001327f.png)
> > Sleeves are same as query image
> > ![image](https://user-images.githubusercontent.com/28961913/92902326-cfd1bc00-f43e-11ea-94d8-b417c21c7018.png)
> > Overlaying the back layer of t-shirt on front neck
> > ![image](https://user-images.githubusercontent.com/28961913/92902379-dbbd7e00-f43e-11ea-83e4-96895e621768.png)
> > Left arms cloth is not overlaying
> > ![image](https://user-images.githubusercontent.com/28961913/92902452-e9730380-f43e-11ea-8dd5-c5a5d2d5cd9e.png)
> > Cloth is same as query image
> > ![image](https://user-images.githubusercontent.com/28961913/92906399-107f0480-f442-11ea-8b1a-f7b8a40f9e63.png)
> 
> May I ask how to print the final result:try-on image? how do you combine them in the last column of the picture？

Hi @Mushahid2521 , The test.py of repo does that.  Find the below line
```
        combine = torch.cat([a[0],d[0],b[0],c[0],rgb[0]], 2).squeeze()
```

**Comment by Mushahid2521** [comment]
@MotiBaadror I am getting edges in the body parsing. I am using this pretrained model. 
https://github.com/PeikeLi/Self-Correction-Human-Parsing trained on LIP dataset. 

Looks like you are getting good enough segmentation.
I get the below result. Will you please let me know which model you are using? 
Thanks a lot.
![sample2](https://user-images.githubusercontent.com/16560161/93739587-7ae63080-fc0a-11ea-84b2-837d741e006a.jpg)

**Comment by MotiBaadror** [comment]
> @MotiBaadror I am getting edges in the body parsing. I am using this pretrained model.
> https://github.com/PeikeLi/Self-Correction-Human-Parsing trained on LIP dataset.
> 
> Looks like you are getting good enough segmentation.
> I get the below result. Will you please let me know which model you are using?
> Thanks a lot.
> ![sample2](https://user-images.githubusercontent.com/16560161/93739587-7ae63080-fc0a-11ea-84b2-837d741e006a.jpg)

Hi @Mushahid2521 , for segmentation, I am using the same model you mentioned.

**Comment by Mushahid2521** [comment]
@MotiBaadror  I think there is a post processing step to follow. I have tried with both CIHP_PGN and the mentioned model. I am getting edges in the segmentation. I have also tried with one of the test image. In that case also I find edges in the segmentation output. Will you please help on this?

**Comment by frankiesquared** [comment]
@MotiBaadror @Mushahid2521 @ChongjianGE @taliegemen @ecarist 
How do I generate a file like those in the folder /DeepFashion_Try_On/Data_preprocessing/test_label ?
I followed your directions to use https://github.com/PeikeLi/Self-Correction-Human-Parsing trained on LIP dataset and got below?

![LIPoutput](https://user-images.githubusercontent.com/59422631/95684611-b3dd4980-0ba7-11eb-87a6-0f61989e5932.png)

The files in  /DeepFashion_Try_On/Data_preprocessing/test_label by default look black with a silhouette.

![000010_0](https://user-images.githubusercontent.com/59422631/95684641-e5561500-0ba7-11eb-8352-b9dc93d67d28.png)

Thanks

**Comment by levindabhi** [comment]
@frankiesquared you can refer my fork of https://github.com/levindabhi/Self-Correction-Human-Parsing-for-ACGPN original repo to generate test_label.

**Comment by ujjawalcse** [comment]
Hi!,
I'm also trying this repository on custom data.
I extracted pose using openpose link and code snippet provided by @MotiBaadror .
And label image are extracted using the above mentioned repository [](https://github.com/levindabhi/Self-Correction-Human-Parsing-for-ACGPN). also I checked after drawing the palette,it's looking fine.
And all other folders required are same as the data given.
But I'm not getting satisfactory result.
You can have a look below 
![1](https://user-images.githubusercontent.com/32773256/104284089-7aceb100-54d7-11eb-9e63-37f3cc8c6908.jpg)
![3](https://user-images.githubusercontent.com/32773256/104284108-84f0af80-54d7-11eb-9293-ccc4eb69c67e.jpg)
![4](https://user-images.githubusercontent.com/32773256/104284116-8ae69080-54d7-11eb-9bef-d4d63fef337d.jpg)
![6](https://user-images.githubusercontent.com/32773256/104284126-9043db00-54d7-11eb-845e-f160c7af3423.jpg)
![7](https://user-images.githubusercontent.com/32773256/104284147-98037f80-54d7-11eb-9c95-893399d92634.jpg)

Please provide any suggestion to improve this result.
Thanks

**Comment by levindabhi** [comment]
@ujjawalcse the main problem I see is the difference in training distribution and testing distribution. Both models (self-correction human parsing model and ACGPN) is trained on a different type of data then you inputting in testing. Retraining on your dataset might help.

**Comment by ujjawalcse** [comment]
Thanks @levindabhi, I will try PGN for segmentation with labels required for ACGPN. And Also, I will try to train with my domain data after creating all the folders required for training.

**Comment by amandazw** [comment]
Hi,I use my own dataset train my model,and preprocessing the data using the methods upper mentioned.But the result are not satisfied.
![000001_0](https://user-images.githubusercontent.com/20120287/119470983-6f3fa080-bd7b-11eb-94b0-6fc8fc2db280.jpg)
![000170_0](https://user-images.githubusercontent.com/20120287/119471014-7797db80-bd7b-11eb-9675-a4dd8a0697f4.jpg)
![000004_0](https://user-images.githubusercontent.com/20120287/119471051-81b9da00-bd7b-11eb-97ed-10c8a0869727.jpg)
Anybody can help me?thx!

**Comment by hamzanaeem1999** [comment]
@Mushahid2521 
Can you please tell me that what is the need of segmentation trained on LIP dataset and the pose of person . please tell me these two questions ?

**Comment by HAN-YANG-HIT** [comment]
Nice discussions. Tryon is a fast developing area, better results will be given by newer papers. Artifacts are somehow inevitable.
